File: frontend/src/utils/formulaValidation.js
/**
 * Detects circular dependencies in account formulas
 *
 * @param {number|null} currentAccountId - ID of account being edited (null for new)
 * @param {Array} proposedFormula - Array of {accountId, coefficient} or {account_id, coefficient}
 * @param {Array} allAccounts - All accounts with their formulas
 * @returns {Object} { hasCircle: boolean, errorMessage: string }
 */
export function detectCircularDependency(currentAccountId, proposedFormula, allAccounts) {
  // Build dependency graph from all accounts
  const graph = new Map();
  const accountNames = new Map();

  for (const account of allAccounts) {
    accountNames.set(account.id, account.account_name);
    if (account.is_calculated && account.formula) {
      const deps = account.formula.map((f) => f.account_id);
      graph.set(account.id, deps);
    } else {
      graph.set(account.id, []);
    }
  }

  // Check each account in the proposed formula
  for (const item of proposedFormula) {
    const depId = item.accountId ?? item.account_id;

    // Self-reference check
    if (depId === currentAccountId) {
      return {
        hasCircle: true,
        errorMessage: 'An account cannot reference itself in its formula',
      };
    }

    // For new accounts (currentAccountId is null), we can't have cycles yet
    if (currentAccountId === null) {
      continue;
    }

    // DFS to detect if depId can reach currentAccountId
    const visited = new Set();
    const path = [];

    if (canReach(depId, currentAccountId, graph, visited, path)) {
      // Build path for error message
      const names = [accountNames.get(currentAccountId), accountNames.get(depId)];
      for (const id of path) {
        names.push(accountNames.get(id));
      }
      names.push(accountNames.get(currentAccountId));
      return {
        hasCircle: true,
        errorMessage: `Circular dependency: ${names.join(' → ')}`,
      };
    }
  }

  return { hasCircle: false, errorMessage: '' };
}

function canReach(from, target, graph, visited, path) {
  if (from === target) {
    return true;
  }
  if (visited.has(from)) {
    return false;
  }
  visited.add(from);

  const deps = graph.get(from) || [];
  for (const dep of deps) {
    if (canReach(dep, target, graph, visited, path)) {
      path.push(dep);
      return true;
    }
  }

  return false;
}

File: backend/internal/validation/formula_validation.go
package validation

import (
	"fmt"

	"finance-tracker/internal/models"
)

// ValidateFormulaForCycles checks if adding/updating a formula would create a circular dependency.
// accountID: The ID of the account being created/updated (0 for new accounts)
// formula: The proposed formula for this account
// allAccounts: All existing accounts in the system
// Returns an error if a cycle is detected, nil otherwise
func ValidateFormulaForCycles(accountID int, formula []models.FormulaItem, allAccounts []models.Account) error {
	// Build dependency graph: accountID -> list of account IDs it depends on
	graph := make(map[int][]int)
	accountNames := make(map[int]string)

	for _, account := range allAccounts {
		accountNames[account.ID] = account.AccountName
		if account.IsCalculated && len(account.Formula) > 0 {
			deps := make([]int, 0, len(account.Formula))
			for _, item := range account.Formula {
				deps = append(deps, item.AccountID)
			}
			graph[account.ID] = deps
		}
	}

	// Check each account in the proposed formula
	for _, item := range formula {
		// Self-reference check
		if item.AccountID == accountID && accountID != 0 {
			return fmt.Errorf("circular dependency: account cannot reference itself")
		}

		// For new accounts (accountID == 0), we only need to check self-reference
		// which isn't possible since we don't have an ID yet
		if accountID == 0 {
			continue
		}

		// DFS to detect if this dependency can reach back to accountID
		visited := make(map[int]bool)
		path := []int{item.AccountID}

		if canReach(item.AccountID, accountID, graph, visited, &path) {
			// Build the cycle path for error message
			cyclePath := accountNames[accountID] + " -> " + accountNames[item.AccountID]
			for i := 0; i < len(path)-1; i++ {
				cyclePath += " -> " + accountNames[path[i]]
			}
			cyclePath += " -> " + accountNames[accountID]
			return fmt.Errorf("circular dependency detected: %s", cyclePath)
		}
	}

	return nil
}

// canReach uses DFS to determine if 'from' can reach 'target' through the dependency graph
func canReach(from, target int, graph map[int][]int, visited map[int]bool, path *[]int) bool {
	if from == target {
		return true
	}
	if visited[from] {
		return false
	}
	visited[from] = true

	deps := graph[from]
	for _, dep := range deps {
		*path = append(*path, dep)
		if canReach(dep, target, graph, visited, path) {
			return true
		}
		*path = (*path)[:len(*path)-1]
	}

	return false
}
